ai_wars/spaceship.py

```diff
@@ -7,72 +7,106 @@
-				(self.surface.get_height() - self.height):
-			self.y -= self.direction.y * self.acceleration
-			self.x -= self.direction.x * self.acceleration
+				# move sprite
+				self.x = valid_pos_x
+				self.y = valid_pos_y
 
-		if action == EnumAction.RIGHT and self.x + self.velocity < \
-				(self.surface.get_width() - self.width):
-			self.rotate(clockwise=True)
+				self.refresh_hitbox_coordinates()
 
-		if action == EnumAction.LEFT and self.x - self.velocity > 0:
-			self.rotate(clockwise=False)
+			case EnumAction.BACKWARD:
+				new_position_x = self.x - self.direction.x * self.acceleration
+				new_position_y = self.y - self.direction.y * self.acceleration
+				# correct the position at the end of an action to stay within the screen bounds
+				valid_pos_x = np.clip(new_position_x, 0, self.screen.get_width())
+				valid_pos_y = np.clip(new_position_y, 0, self.screen.get_height())
 
-	def shoot(self) -> None:
+				self.x = valid_pos_x
+				self.y = valid_pos_y
+
+				self.refresh_hitbox_coordinates()
+
+			case EnumAction.RIGHT:
+				self._rotate(clockwise=True)
+
+			case EnumAction.LEFT:
+				self._rotate(clockwise=False)
+
+			case EnumAction.SHOOT:
+				self._shoot()
+
+	def _shoot(self) -> None:
 		"""public method to create a bullet and append it to the bullet list"""
 		bullet_velocity = self.direction * 3
-		bullet = Bullet(self.x+np.floor(self.width/2)-5, self.y, self.height, self.width, \
-						load_sprite("ai_wars/img/bullet.png"), bullet_velocity)
+		#TODO We are passing here the wrong height and width (that of the ship), the bullet class
+		# can get it itself using the img
+		bullet = Bullet(self.x, self.y - np.floor(self.height/2), self.height, self.width, \
+						load_sprite("ai_wars/img/bullet.png"), bullet_velocity, self)
 		self.bullet_append(bullet)
 
-	def got_hit(self) -> None:
-		"""public method to decrease the score, everytime the ship got hit"""
-		self.score -= 10
-
-	def rotate(self, clockwise=True) -> None:
+	def _rotate(self, clockwise: bool) -> None:
 		"""public method to rotate the ship in clockwise direction"""
 		# rotate the direction vector
 		sign = 1 if clockwise else -1
 		angle = 3 * sign
 		self.direction.rotate_ip(angle)
 
-	def draw(self, surface: pygame.Surface) -> None:
-		# player_rec = pygame.Rect(self.x, self.y, self.height, self.width)
-		# surface.blit(self.sprite, (self.x, self.y))
+	def draw(self, screen: pygame.Surface) -> None:
+		"""public method to draw the rotated version of the spaceship"""
 		angle = self.direction.angle_to(UP)
 		rotated_surface = pygame.transform.rotozoom(self.sprite, angle, 1.0)
 		rotated_surface_size = Vector2(rotated_surface.get_size())
 		blit_position = Vector2(self.x, self.y) - rotated_surface_size * 0.5
-		surface.blit(rotated_surface, blit_position)
+		screen.blit(rotated_surface, blit_position)
+		
+		#Debugging - Draw Hitbox
+		#pygame.draw.rect(screen, (0,255,0), self.hitbox)
+
+	def update(self, subject: Subject) -> None:
```

Comment:
Eigentlich brauchen wir die Methode und Observer-Klassen nicht mehr oder? Dem SHip ist es ja eigentlich egal, ob der score sich ändert

Reply:
jo, entferne ich in dem Observer Branch und merge das dann auf Main

Reply:
alles klar